### DISCUSSION TITLE
Quantifiers with recursive datatypes causes 'incomplete'

### DISCUSSION THREAD

**Discussion by aidanhs** [cvc5/cvc5#10742] [Q&A] [answered]
I'm new to SMT solvers and have bounced off z3 and yices because of limitations with quantifiers. cvc5 is better, but I'm struggling with this `incomplete`:

```
(set-logic ALL)
(set-option :produce-models true)
(set-option :produce-proofs true)
(set-option :finite-model-find true)

(declare-datatypes ((Ty 0))
    ((
        (U32)
        (foo)
        (identity (inner Ty))
    ))
)

(declare-fun rets (Ty) (Set Ty))
(define-fun path ((ret Ty)) Bool
    (exists ((f Ty))
        (set.member ret (rets f))
    )
)

(assert (= (rets U32) (set.singleton U32)))
(assert (= (rets foo) (set.singleton U32)))
(assert (forall ((inner Ty))
    (= (rets (identity inner)) (set.singleton inner))
))

(assert (path U32))
(check-sat)
(get-model)
```

This returns `incomplete`, but commenting out all lines relating to `identity` returns `sat`. It feels like SMT solvers typically don't advertise themselves as being the first choice for quantified formulas so perhaps I need to look at other classes of tools? I get the impression I'm going to need a deeper understanding of SMT solver strategies to debug issues like this if I continue on this path.

Advice appreciated!

**Comment by ajreynol** [comment, marked as answer]
For benchmarks that are "sat" and involve quantified formulas, SMT solvers will by default only do well if the interpretation of functions in models are "almost constant", i.e. a table of table of I/O pairs, followed by a default value.

In your example, `rets` has a non-trivial interpretation, so traditional SMT techniques won't work if you want "sat".

For this benchmark, the best chance you have of answering "sat" is to use `--sygus-inference=on`, which will recast the input problem as a syntax-guided synthesis problem.  You can further use `-o sygus` to see the candidate interpretations being tried.

Commenting out the set-option commands, cvc5 is able to find the required interpretation of `rets` after 2 minutes or so:

```
% cvc5 discussion10742.smt2 --sygus-inference=on 
sat
(
(define-fun rets ((arg0 Ty)) (Set Ty) (ite ((_ is identity) arg0) (set.singleton (inner arg0)) (set.singleton U32)))
)
```

In general, `--sygus-inference=on` invokes a "heavy hammer" which may be slow in general since it based on enumerating all possible interpretations over the grammar of the background theory, but can lead to finding non-trivial models.

You could try to rephrase the problem as SyGuS and then provide an expected grammar of what you expect the model for `rets` to look like, which can lead to better performance.  See https://sygus-org.github.io/language/ for the SyGuS input language, which is similar to smt2 and cvc5 supports.
